nearcore/src/config_validate.rs

```diff
@@ -0,0 +1,140 @@
+use near_config_utils::{ValidationError, ValidationErrors};
+
+use crate::config::Config;
+
+/// Validate Config extracted from config.json.
+/// This function does not panic. It returns the error if any validation fails.
+pub fn validate_config(config: &Config) -> Result<(), ValidationError> {
+    let mut validation_errors = ValidationErrors::new();
+    let mut config_validator = ConfigValidator::new(config, &mut validation_errors);
+    tracing::info!(target: "config", "Validating Config, extracted from config.json...");
+    config_validator.validate()
+}
+
+struct ConfigValidator<'a> {
+    config: &'a Config,
+    validation_errors: &'a mut ValidationErrors,
+}
+
+impl<'a> ConfigValidator<'a> {
+    fn new(config: &'a Config, validation_errors: &'a mut ValidationErrors) -> Self {
+        Self { config, validation_errors }
+    }
+
+    fn validate(&mut self) -> Result<(), ValidationError> {
+        self.validate_all_conditions();
+        self.result_with_full_error()
+    }
+
+    /// this function would check all conditions, and add all error messages to ConfigValidator.errors
+    fn validate_all_conditions(&mut self) {
+        if self.config.archive == false && self.config.save_trie_changes == Some(false) {
+            let error_message = format!("Configuration with archive = false and save_trie_changes = false is not supported because non-archival nodes must save trie changes in order to do do garbage collection.");
+            self.validation_errors
+                .push_errors(ValidationError::ConfigSemanticsError { error_message: error_message })
+        }
+
+        if self.config.consensus.min_block_production_delay
+            > self.config.consensus.max_block_production_delay
+        {
+            let error_message = format!(
+                "min_block_production_delay: {:?} is greater than max_block_production_delay: {:?}",
+                self.config.consensus.min_block_production_delay,
+                self.config.consensus.max_block_production_delay
+            );
+            self.validation_errors
+                .push_errors(ValidationError::ConfigSemanticsError { error_message: error_message })
+        }
+
+        if self.config.consensus.min_block_production_delay
+            > self.config.consensus.max_block_wait_delay
+        {
+            let error_message = format!(
+                "min_block_production_delay: {:?} is greater than max_block_wait_delay: {:?}",
+                self.config.consensus.min_block_production_delay,
+                self.config.consensus.max_block_wait_delay
+            );
+            self.validation_errors
+                .push_errors(ValidationError::ConfigSemanticsError { error_message: error_message })
+        }
+
+        if self.config.consensus.sync_height_threshold == 0 {
```

Comment:
does this need to be enforced? I agree that it would maybe be a bit strange to set it to zero. But I don't quite understand why it should be treated as actually incorrect